src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs

```diff
@@ -184,64 +190,93 @@ private static void AdjustSpans(ArrayBuilder<ClassifiedSpan> spans, TextSpan wid
+                        syntacticPart.TextSpan.Start,
+                        firstSemanticPart.TextSpan.Start)));
                 }
-                // We might already have a semantic part for the given TextSpan, in
-                // which case we don't want to add the syntactic part unless it's an
-                // additive type name (e.g. `static`).
-                else if (finalParts.Count == 0 ||
-                    !finalParts[^1].TextSpan.Equals(syntaxPartAndSpan.TextSpan) ||
-                    ClassificationTypeNames.AdditiveTypeNames.Contains(syntaxPartAndSpan.ClassificationType))
+
+                for (var i = 0; i < tempBuffer.Count - 1; i++)
                 {
-                    finalParts.Add(syntaxPartAndSpan);
+                    var semanticPart1 = tempBuffer[i];
+                    var semanticPart2 = tempBuffer[i + 1];
+
+                    Debug.Assert(semanticPart1.TextSpan.OverlapsWith(syntacticPart.TextSpan));
+                    Debug.Assert(semanticPart1.TextSpan.OverlapsWith(syntacticPart.TextSpan));
+
+                    if (semanticPart1.TextSpan.End < semanticPart2.TextSpan.Start)
+                    {
+                        finalParts.Add(new ClassifiedSpan(syntacticPart.ClassificationType, TextSpan.FromBounds(
+                            semanticPart1.TextSpan.End,
+                            semanticPart2.TextSpan.Start)));
+                    }
                 }
 
-                if (replacementIndex >= 0)
+                if (lastSemanticPart.TextSpan.End < syntacticPart.TextSpan.End)
                 {
-                    // There may be multiple semantic parts corresponding to a single
-                    // syntactic part, so we might need to go through a syntactic part
-                    // multiple times to verify. For example, this is the case with
-                    // verbatim string literals containing string escape characters.
-                    i--;
-
-                    // If we found a semantic replacement, update the lastIndex.
-                    // That way we can start searching from that point instead 
-                    // of checking all the elements each time.
-                    lastReplacementIndex = replacementIndex + 1;
+                    finalParts.Add(new ClassifiedSpan(syntacticPart.ClassificationType, TextSpan.FromBounds(
+                        lastSemanticPart.TextSpan.End,
+                        syntacticPart.TextSpan.End)));
                 }
             }
 
-            return finalParts.ToImmutable();
-        }
-
-        private static bool IsClassifiedAsText(ClassifiedSpan partAndSpan)
-        {
-            // Don't take 'text' from the semantic parts.  We'll get those for the 
-            // spaces between the actual interesting semantic spans, and we don't 
-            // want them to override actual good syntax spans.
-            return partAndSpan.ClassificationType == ClassificationTypeNames.Text;
```

Comment:
does this not apply anymore?

Reply:
it does not, due to us only filling in the gaps *after* syntactic and semantic tokens are merged.  This was previously necessary because we'd fill the gaps in both first, *then* try to merge.

Reply:
ahh of course